python/ql/src/experimental/semmle/python/frameworks/NoSQL.qll

```diff
@@ -0,0 +1,187 @@
+  private DataFlow::LocalSourceNode mongoClientCollection(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      exists(SubscriptNode subscript | result.asCfgNode() = subscript |
+        subscript.getObject() = mongoClientDB().asCfgNode()
+      )
+      or
+      result.(DataFlow::AttrRead).getObject() = mongoClientDB()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = mongoClientCollection(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `MongoClient` collection. */
+  private DataFlow::Node mongoClientCollection() {
+    mongoClientCollection(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** This class represents names of find_* relevant MongoClient collection level operation methods. */
+  private class MongoClientMethodNames extends string {
+    MongoClientMethodNames() {
+      // the find_one_or_404 method is only found in the Pymongo Flask library.
+      this in [
+          "find", "find_raw_batches", "find_one", "find_one_and_delete", "find_and_modify",
+          "find_one_and_replace", "find_one_and_update", "find_one_or_404"
+        ]
+    }
+  }
+
+  /** Gets a reference to a `MongoClient` Collection method. */
+  private DataFlow::Node mongoClientMethod() {
+    result.(DataFlow::AttrRead).getAttributeName() instanceof MongoClientMethodNames and
+    (
+      result.(DataFlow::AttrRead).getObject() = mongoClientCollection() or
+      result.(DataFlow::AttrRead) = mongoClientCollection()
+    )
+  }
+
+  /** Gets a reference to a `MongoClient` call */
+  private class MongoClientCall extends DataFlow::CallCfgNode, NoSQLQuery::Range {
+    MongoClientCall() { this.getFunction() = mongoClientMethod() }
+
+    override DataFlow::Node getQuery() { result = this.getArg(0) }
+  }
+
+  private class MongoEngineObjectsCall extends DataFlow::CallCfgNode, NoSQLQuery::Range {
+    MongoEngineObjectsCall() {
+      this =
+        API::moduleImport("mongoengine")
+            .getMember("Document")
+            .getASubclass()
+            .getMember("objects")
+            .getACall()
+    }
+
+    override DataFlow::Node getQuery() { result = this.getArgByName(_) }
+  }
+
+  private class FlaskMongoEngineObjectsCall extends DataFlow::CallCfgNode, NoSQLQuery::Range {
+    FlaskMongoEngineObjectsCall() {
+      this =
+        API::moduleImport("flask_mongoengine")
+            .getMember("MongoEngine")
+            .getReturn()
+            .getMember("Document")
+            .getASubclass()
+            .getMember("objects")
+            .getACall()
+    }
+
+    override DataFlow::Node getQuery() { result = this.getArgByName(_) }
+  }
+
+  private DataFlow::Node flaskMongoEngineInstance() {
+    result = API::moduleImport("flask_mongoengine").getMember("MongoEngine").getReturn().getAUse()
+  }
+
+  /**
+   * A MongoEngine.Document subclass which represents a single MongoDB table.
+   */
+  private class FlaskMongoEngineDocumentClass extends ClassValue {
```

Comment:
Likely need to update this code to also cover `EmbeddedDocument`. This likely requires a change in naming here.

Reply:
done